rfcs/0051-stop-transaction-processing.md

```diff
@@ -0,0 +1,146 @@
+# Stop processing transactions after a certain slot
+
+This PR describes the feature to stop processing transactions after a certain
+slot, to be used in the Berkeley hard fork.
+
+## Summary
+
+Transactions come from a client or the gossip network and are processed by BPs
+and SNARK workers to be included in blocks.
+
+In this RFC, the procedure to stop processing any new transactions after a
+certain slot is described. This is, any blocks produced after that slot will
+include no transaction at all.
+
+## Motivation
+
+In a hard fork scenario, we want to halt the preceding network and produce a new
+genesis ledger for the succeeding network. We want this to happen by defining a
+point in time (slot) where the network stops accepting transactions a keeps
+producing blocks free of transactions until it reaches a consensus state.
+
+This feature enables part of this procedure, by allowing users to define a slot
+at which the node will stop accepting any new transactions or including them in
+blocks (in case it is a BP node).
+
+## Detailed design
+
+The procedure to stop processing transactions after a certain slot will act on
+different components of the node, namely, the CLI, the transaction pool, the
+snark pool, and the block producer.
+and the CLI. The procedure will be as follows:
+
+* The CLI will allow a slot at which the node will stop processing transactions
+  to be set. This will be done by setting a new optional flag `--slot-tx-end`
+  that will be used by the transaction pool, snark pool, and the block producer
+  to stop processing transactions after a certain slot. If the flag is not set,
+  the node will continue processing transactions as usual.
+* The node will stop accepting new transactions from the client after the slot
+  set by the CLI flag.
+* The transaction pool will stop accepting new transactions after the slot set
+  by the CLI flag.
+* The transaction pool will drop all transactions from the pool after the slot
```

Comment:
Is this necessary? The transactions will still be applicable on the post-HF network; surely it would make more sense (at least, for block producers) to keep them and (potentially) port them over?

Reply:
I would strongly advise against making changes to the transaction pool to drop existing transactions and to no longer accept new transactions. In order to reduce the testing surface area this close to the hard fork, we should limit the necessary changes to the daemon for this transaction stop logic. The only necessary rule here is to stop including transactions into blocks -- whether or not the node stops accepting transactions over gossip, or is retaining transactions in the mempool that it does not intend to apply, are unimportant.

Reply:
I think we should purge the mempool, and stop new transactions from entering it, for two reasons:

1. If the transactions are in the mempool, users might expect them to be included on the chain at some point. But without additional work (and additional testing), they will not. The user experience will be much cleaner if transactions that do not make it to the chain are not expected to be included, and will need to be resubmitted after the hard fork.
2. While there is a risk that we might break something with the change to the mempool, there is also a risk that we might break something by doing an incomplete change. If we only stop transactions from entering blocks, they will accumulate in the mempool, and the mempools will get quite full. That could arm a lingering performance hazard, and if we do not do this change, then we will need to test for that. All in all, I think it is less risky to clear the mempool, than to not change it and have it fill up with transactions.

Reply:
I think we shouldn't, at least in part because I still believe it adds needless complexity to no benefit.

To your specific concerns:
1. Users may have similar expectations around their transactions once they're sent. I'm not sure I believe that their transactions disappearing is a 'cleaner' UX than them sitting idle in the pool and then the network going offline, if they aren't aware of the fork.
2. The mempools already have mechanisms for dealing with being full, and reasonable limits to prevent them from causing performance problems. We have extensive real-world testing of this behaviour from mainnet, where conditions triggered exactly the same scenario through over-capacity demand instead, and from occasional DOS attempts. The risky thing, to me, seems to change systems unnecessarily.